Firestore/core/src/local/leveldb_remote_document_cache.cc

```diff
@@ -170,115 +171,82 @@ MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
 }
 
 MutableDocumentMap LevelDbRemoteDocumentCache::GetAllExisting(
-    const DocumentKeySet& keys) {
-  MutableDocumentMap docs = LevelDbRemoteDocumentCache::GetAll(keys);
+    const DocumentVersionMap key_version_map) const {
   MutableDocumentMap result;
-  for (const auto& kv : docs) {
-    const DocumentKey& key = kv.first;
-    auto& document = kv.second;
+  for (const auto& pair : key_version_map) {
+    auto document = Get(pair.first).WithReadTime(pair.second);
     if (document.is_found_document()) {
-      result = result.insert(key, document);
+      result = result.insert(pair.first, std::move(document));
     }
   }
-
   return result;
 }
 
-model::MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
+MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
     const std::string& collection_group,
     const model::IndexOffset& offset,
     size_t limit) const {
-  (void)collection_group;
-  (void)offset;
-  (void)limit;
-  // TODO(cheryllin): Implement GetAll() together with backfiller
-  return model::MutableDocumentMap();
-}
+  HARD_ASSERT(limit > 0u, "Limit should be at least 1");
+  const auto parents = index_manager_->GetCollectionParents(collection_group);
+  std::vector<ResourcePath> collections;
+  collections.reserve(parents.size());
+  for (const auto& parent : parents) {
+    collections.push_back(parent.Append(collection_group));
+  }
 
-MutableDocumentMap LevelDbRemoteDocumentCache::GetAll(
-    const model::ResourcePath& path, const model::IndexOffset& offset) {
-  // Use the query path as a prefix for testing if a document matches the query.
-  size_t immediate_children_path_length = path.size() + 1;
-
-  if (offset.read_time() != SnapshotVersion::None()) {
-    // Execute an index-free query and filter by read time. This is safe since
-    // all document changes to queries that have a
-    // last_limbo_free_snapshot_version (`since_read_time`) have a read time
-    // set.
-    std::string start_key =
-        LevelDbRemoteDocumentReadTimeKey::KeyPrefix(path, offset.read_time());
-    auto it = db_->current_transaction()->NewIterator();
-    it->Seek(util::ImmediateSuccessor(start_key));
-
-    DocumentKeySet remote_keys;
-
-    LevelDbRemoteDocumentReadTimeKey current_key;
-    for (; it->Valid() && current_key.Decode(it->key()); it->Next()) {
-      const ResourcePath& collection_path = current_key.collection_path();
-      if (collection_path != path) {
+  MutableDocumentMap result;
+  for (auto path = collections.cbegin();
+       path != collections.cend() && result.size() < limit; path++) {
+    const auto map = GetAll(*path, offset);
+    for (const auto& pair : map) {
+      result = result.insert(pair.first, pair.second);
+      if (result.size() == limit) {
```

Comment:
Unlike other platforms (which iterates over resultset), here you call `GetAll` for all documents, then throw away some if the result is over the limit.

You can pass `limit` down to `GetAll(*path, offset)` and make sure we only process enough documents.